lib/rubocop/cop/rspec/rails/inferred_spec_type.rb

```diff
@@ -0,0 +1,127 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      module Rails
+        # Identifies redundant spec type.
+        #
+        # @safety
+        #   This cop is marked as unsafe because
+        #   `config.infer_spec_type_from_file_location!` may not be enabled.
+        #
+        # @example
+        #
+        #   # bad
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User, type: :model
+        #
+        #   # good
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User
+        #
+        #   # good
+        #   # spec/models/user_spec.rb
+        #   RSpec.describe User, type: :request
+        #
+        class InferredSpecType < Base
+          extend AutoCorrector
+
+          # @return [Array<Hash>]
+          INFERENCES = [
+            { prefix: 'spec/channels/', type: :channel },
+            { prefix: 'spec/controllers/', type: :controller },
+            { prefix: 'spec/features/', type: :feature },
+            { prefix: 'spec/generator/', type: :generator },
+            { prefix: 'spec/helpers/', type: :helper },
+            { prefix: 'spec/jobs/', type: :job },
+            { prefix: 'spec/mailboxes/', type: :mailbox },
+            { prefix: 'spec/mailers/', type: :mailer },
+            { prefix: 'spec/models/', type: :model },
+            { prefix: 'spec/requests/', type: :request },
+            { prefix: 'spec/integration/', type: :request },
+            { prefix: 'spec/api/', type: :request },
+            { prefix: 'spec/routing/', type: :routing },
+            { prefix: 'spec/system/', type: :system },
+            { prefix: 'spec/views/', type: :view }
+          ].freeze
+
+          MSG = 'Remove redundant spec type.'
+
+          RESTRICT_ON_SEND = %i[
+            describe
+          ].freeze
+
+          # @param [RuboCop::AST::SendNode] node
+          def on_send(node)
```

Comment:
Is it sufficient to only check ones that are passed to `on_top_level_group`?

Reply:
`#on_top_level_example_group` seems to solve several problems, but I'm stuck with a problem that autocorrect not working properly when using it. It seems that the node obtained from `#on_top_level_example_group` refers to a different buffer instance from the corrector expects. Is there any workaround for this?

```diff
diff --git a/lib/rubocop/cop/rspec/rails/inferred_spec_type.rb b/lib/rubocop/cop/rspec/rails/inferred_spec_type.rb
index 56f3005..374b759 100644
--- a/lib/rubocop/cop/rspec/rails/inferred_spec_type.rb
+++ b/lib/rubocop/cop/rspec/rails/inferred_spec_type.rb
@@ -49,14 +49,12 @@ module RuboCop
         class InferredSpecType < Base
           extend AutoCorrector
 
-          MSG = 'Remove redundant spec type.'
+          include TopLevelGroup
 
-          RESTRICT_ON_SEND = %i[
-            describe
-          ].freeze
+          MSG = 'Remove redundant spec type.'
 
-          # @param [RuboCop::AST::SendNode] node
-          def on_send(node)
+          # @param [RuboCop::AST::BlockNode] node
+          def on_top_level_example_group(node)
             pair_node = describe_with_type(node)
             return unless pair_node
             return unless inferred_type?(pair_node)
@@ -70,17 +68,20 @@ module RuboCop
           private
 
           # @!method describe_with_type(node)
-          #   @param [RuboCop::AST::SendNode] node
+          #   @param [RuboCop::AST::BlockNode] node
           #   @return [RuboCop::AST::PairNode, nil]
           def_node_matcher :describe_with_type, <<~PATTERN
-            (send
-              { (const nil? :RSpec) | nil? }
-              _
-              _
-              _*
-              ({ hash | kwargs }
-                <$(pair (sym :type) (sym _)) ...>
+            (block
+              (send
+                { (const nil? :RSpec) | nil? }
+                _
+                _
+                _*
+                ({ hash | kwargs }
+                  <$(pair (sym :type) (sym _)) ...>
+                )
               )
+              ...
             )
           PATTERN
 ```

```
$ bundle exec rspec spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb

Randomized with seed 2324

RuboCop::Cop::RSpec::Rails::InferredSpecType
  with necessary type in keyword arguments
    does not register any offense
  with redundant type and other Hash metadata
    autocorrects offense (FAILED - 1)
  with redundant type in Hash arguments
    autocorrects offense (FAILED - 2)
  with redundant type and receiver-less describe
    autocorrects offense (FAILED - 3)
  with redundant type and other Symbol metadata
    autocorrects offense (FAILED - 4)
  with Inferrences configuration
    autocorrects offense (FAILED - 5)
  with redundant type in keyword arguments
    autocorrects offense (FAILED - 6)
  with default configuration
    does not crash on smoke test: spec/smoke_tests/factory_bot_spec.rb
    does not crash on smoke test: spec/smoke_tests/empty_spec.rb
    does not crash on smoke test: spec/smoke_tests/weird_rspec_spec.rb
    does not crash on smoke test: spec/smoke_tests/no_tests_spec.rb

Failures:

  1) RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and other Hash metadata autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4240 name:"/path/to/project/spec/models/user_spec.rb" is not current 4260 name:"/path/to/project/spec/models/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:51:in `block (3 levels) in <top (required)>'

  2) RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type in Hash arguments autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4280 name:"/path/to/project/spec/models/user_spec.rb" is not current 4300 name:"/path/to/project/spec/models/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:36:in `block (3 levels) in <top (required)>'

  3) RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and receiver-less describe autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4320 name:"/path/to/project/spec/models/user_spec.rb" is not current 4340 name:"/path/to/project/spec/models/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:81:in `block (3 levels) in <top (required)>'

  4) RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and other Symbol metadata autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4360 name:"/path/to/project/spec/models/user_spec.rb" is not current 4380 name:"/path/to/project/spec/models/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:66:in `block (3 levels) in <top (required)>'

  5) RuboCop::Cop::RSpec::Rails::InferredSpecType with Inferrences configuration autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4400 name:"/path/to/project/spec/services/user_spec.rb" is not current 4420 name:"/path/to/project/spec/services/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:104:in `block (3 levels) in <top (required)>'

  6) RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type in keyword arguments autocorrects offense
     Failure/Error:
       corrector.remove(
         node.location.expression.with(
           begin_pos: node.left_sibling.location.expression.end_pos
         )
       )
     
     RuntimeError:
       Correction target buffer 4440 name:"/path/to/project/spec/models/user_spec.rb" is not current 4460 name:"/path/to/project/spec/models/user_spec.rb" under investigation
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:91:in `autocorrect'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:64:in `block in on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/rails/inferred_spec_type.rb:63:in `on_top_level_example_group'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:14:in `block in on_new_investigation'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `each'
     # ./lib/rubocop/cop/rspec/mixin/top_level_group.rb:13:in `on_new_investigation'
     # ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:21:in `block (3 levels) in <top (required)>'

Finished in 0.06853 seconds (files took 0.58951 seconds to load)
11 examples, 6 failures

Failed examples:

rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:44 # RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and other Hash metadata autocorrects offense
rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:29 # RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type in Hash arguments autocorrects offense
rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:74 # RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and receiver-less describe autocorrects offense
rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:59 # RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type and other Symbol metadata autocorrects offense
rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:97 # RuboCop::Cop::RSpec::Rails::InferredSpecType with Inferrences configuration autocorrects offense
rspec ./spec/rubocop/cop/rspec/rails/inferred_spec_type_spec.rb:14 # RuboCop::Cop::RSpec::Rails::InferredSpecType with redundant type in keyword arguments autocorrects offense

Randomized with seed 2324
```

Reply:
```ruby
        RSpec.describe User do
          describe 'inner', type: :model do
```
Makes total sense to me 👍 
This is a sufficient argument for me to drop the idea to only check top-level example groups.